File: solutions/go/days/2.go
package days

import (
	"fmt"
	"strconv"
	"strings"
)

type Day2 struct{}

func (d Day2) Part1(input []string) string {
	total := 0
	redLimit := 12
	greenLimit := 13
	blueLimit := 14

gameLoop:
	for _, line := range input {
		splitLine := strings.Split(line, ":")
		id, _ := strconv.Atoi(splitLine[0][5:])

		gameString := strings.Replace(splitLine[1], " ", "", -1)
		roundStrings := strings.Split(gameString, ";")
		for _, roundString := range roundStrings {
			colorStrings := strings.Split(roundString, ",")
			for _, colorString := range colorStrings {
				if colorString[len(colorString)-3:] == "red" {
					redCount, _ := strconv.Atoi(colorString[:len(colorString)-3])
					if redCount > redLimit {
						continue gameLoop
					}
				}
				if len(colorString) > 4 && colorString[len(colorString)-4:] == "blue" {
					blueCount, _ := strconv.Atoi(colorString[:len(colorString)-4])
					if blueCount > blueLimit {
						continue gameLoop
					}
				}
				if len(colorString) > 5 && colorString[len(colorString)-5:] == "green" {
					greenCount, _ := strconv.Atoi(colorString[:len(colorString)-5])
					if greenCount > greenLimit {
						continue gameLoop
					}
				}
			}
		}
		total += id
	}
	return fmt.Sprintf("%d", total)
}

func (d Day2) Part2(input []string) string {
	total := 0
	for _, line := range input {
		redMax := 0
		blueMax := 0
		greenMax := 0

		splitLine := strings.Split(line, ":")
		gameString := strings.Replace(splitLine[1], " ", "", -1)
		roundStrings := strings.Split(gameString, ";")
		for _, roundString := range roundStrings {
			colorStrings := strings.Split(roundString, ",")
			for _, colorString := range colorStrings {
				if colorString[len(colorString)-3:] == "red" {
					redCount, _ := strconv.Atoi(colorString[:len(colorString)-3])
					if redCount > redMax {
						redMax = redCount
					}
				}
				if len(colorString) > 4 && colorString[len(colorString)-4:] == "blue" {
					blueCount, _ := strconv.Atoi(colorString[:len(colorString)-4])
					if blueCount > blueMax {
						blueMax = blueCount
					}
				}
				if len(colorString) > 5 && colorString[len(colorString)-5:] == "green" {
					greenCount, _ := strconv.Atoi(colorString[:len(colorString)-5])
					if greenCount > greenMax {
						greenMax = greenCount
					}
				}
			}
		}
		total += redMax * greenMax * blueMax
	}
	return fmt.Sprintf("%d", total)
}

func NewDay2() Day2 {
	return Day2{}
}

File: solutions/typescript/days/2.ts
import { Run, Day } from "../common";

const Day2: Day = {
  Part1: (input: Array<string>): string => {
    return input
      .reduce((accumulator: number, line: string) => {
        const game = line.split(": ");
        const id = parseInt(game[0].slice(4));
        const rounds = game[1].split("; ");
        for (let i = 0; i < rounds.length; i++) {
          const colorPairs = rounds[i].split(", ");
          for (let j = 0; j < colorPairs.length; j++) {
            const colorTuple = colorPairs[j].split(" ");
            if (
              (colorTuple[1] == "red" && parseInt(colorTuple[0]) > 12) ||
              (colorTuple[1] == "blue" && parseInt(colorTuple[0]) > 14) ||
              (colorTuple[1] == "green" && parseInt(colorTuple[0]) > 13)
            ) {
              return accumulator;
            }
          }
        }
        return accumulator + id;
      }, 0)
      .toString(10);
  },
  Part2: (input: Array<string>): string => {
    return input
      .reduce((accumulator: number, line: string) => {
        const game = line.split(": ");
        const rounds = game[1].split("; ");
        let redMax = 0,
          blueMax = 0,
          greenMax = 0;
        for (let i = 0; i < rounds.length; i++) {
          const colorPairs = rounds[i].split(", ");
          for (let j = 0; j < colorPairs.length; j++) {
            const colorTuple = colorPairs[j].split(" ");
            const value = parseInt(colorTuple[0]);
            if (colorTuple[1] == "red" && value > redMax) {
              redMax = value;
            }
            if (colorTuple[1] == "blue" && value > blueMax) {
              blueMax = value;
            }
            if (colorTuple[1] == "green" && value > greenMax) {
              greenMax = value;
            }
          }
        }
        return accumulator + redMax * greenMax * blueMax;
      }, 0)
      .toString(10);
  },
};

Run(Day2);
